src/charm.py

```diff
@@ -13,91 +13,189 @@
+
+        self._stored.backend_store_uri = db_uri
+        self._on_config_changed(event)
+
+    def _object_storage_relation_changed(self, event: RelationEvent):
+        """Handle minio relation event."""
+        if not self.unit.is_leader():
+            return
+
+        minio_secrets = yaml.safe_load(event.relation.data.get(event.app, {}).get("data", "{}"))
+        minio_service = minio_secrets.get("service", None)
+
+        if minio_service:
+            minio_url = f"http://{minio_secrets['service']}:{minio_secrets['port']}"
+            self._stored.minio_environment.update({
+                "MLFLOW_S3_ENDPOINT_URL": minio_url,
+                "AWS_ACCESS_KEY_ID": minio_secrets["access-key"],
+                "AWS_SECRET_ACCESS_KEY": minio_secrets["secret-key"],
+            })
+            artifact_root = "s3://mlflow/"
+        else:
+            self._stored.minio_environment.clean()
+            artifact_root = DEFAULT_ARTIFACT_ROOT
+
+        self._stored.artifact_root = artifact_root
+        self._on_config_changed(event)
+
+    def _mlflow_layer(self):
+        """Returns a Pebble configuration layer for Mlflow."""
+        host = self.config["host"]
+        port = self.config["port"]
+        backend_store_uri = self._stored.backend_store_uri
+        artifact_root = self._stored.artifact_root
+        environment = self._stored.minio_environment
+
+        return {
+            "summary": "MLflow server layer",
+            "description": "pebble config layer for MLflow server",
             "services": {
-                "httpbin": {
+                "server": {
                     "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
+                    "summary": "MLflow server",
+                    "command": "mlflow server"
+                               f" --host {host}"
+                               f" --port {port}"
+                               f" --backend-store-uri {backend_store_uri}"
+                               f" --default-artifact-root {artifact_root}",
                     "startup": "enabled",
-                    "environment": {"thing": self.model.config["thing"]},
+                    "environment": environment
                 }
-            },
+            }
         }
-        # Add intial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", pebble_layer, combine=True)
-        # Autostart any services that were defined with startup: enabled
-        container.autostart()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
+
+    def _manage_server_layer(self):
+        """Manage MLflow server layer w/ Pebble."""
+        container = self.unit.get_container("server")
+        mlflow_layer = self._mlflow_layer()
+        services = container.get_plan().to_dict().get("services", {})
```

Comment:
This type of call uses the Pebble API, so I think `except (APIError, ConnectionError) as error` should be caught.

FWIW, `from ops.pebble import APIError, ConnectionError, Layer, ServiceStatus`

Reply:
Yes, but also `container.get_service`, `container.start`, `container.stop` (used in multiple function) and `container.add_layer`. I don't think if I don't do it in the whole code, it makes sense to do it here.

Reply:
From my point of view, pebble issues is not a user problem, but an infrastructure problem. Besides, we are aware as developers that the Pebble API may not be available transiently. If we don't catch any error, the user will see a charm error in "juju status". If that user is lucky, `juju resolve` will fix the problem, but there is not much the user can do. If I'm wrong, and such pebble error could be troubleshooted, then I'm ok leaving it as is.

At this stage, adding a comment mentioning the need to catch possible transient pebble api errors would be enough.

Reply:
What do you think about this approach:
```python
try:
    self._manage_server_layer()
except (ops.pebble.TimeoutError, ops.pebble.ConnectionError):
    self.unit.status = BlockedStatus("Pebble API connection problem.")
    event.defer()  # only in `_on_server_pebble_ready`
    return
```
I tested it manually with broken pebble API.